File: leetcodesln/SearchInRotatedSortedArray.cs
namespace leetcodesln
{
    public class SearchInRotatedSortedArray
    {
        public int Search(int[] nums, int target)
        {
            int l = 0, r = nums.Length - 1;

            while (l < r)
            {
                var mid = l + (r - l) / 2;
                if (nums[mid] < nums[r]) r = mid;
                else l = mid + 1;
            }

            int shuffle = l;
            int low = 0, high = nums.Length - 1;
            while (low <= high)
            {
                var mid = low + (high - low) / 2;
                var realmid = (mid + shuffle) % nums.Length;
                if (nums[realmid] == target) return realmid;
                else if (nums[realmid] > target) high = mid - 1;
                else low = mid + 1;
            }
            return -1;
        }
    }
}

File: Go/questions/SearchInRotatedSortedArray.go
package questions

func SearchInRotatedSortedArray(nums []int, target int) int {
	l, r := 0, len(nums)-1

	for l < r {
		mid := (l + r) >> 1
		if nums[mid] < nums[r] {
			r = mid
		} else {
			l = mid + 1
		}
	}

	shuffle := l

	l, r = 0, len(nums)-1

	for l <= r {
		mid := (l + r) >> 1
		realmid := (mid + shuffle) % len(nums)
		if nums[realmid] == target {
			return realmid
		} else if nums[realmid] > target {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}
	return -1
}
